File: haskell/AdventOfCode/Year2021/Day01.hs
module AdventOfCode.Year2021.Day01 (day01a, day01b) where

import Data.List (tails)

strToIntList :: String -> [Int]
strToIntList string = map (read :: String -> Int) (lines string)

hasIncreased :: [Int] -> [Bool]
hasIncreased lst = zipWith (<) lst (tail lst)

numTrue :: [Bool] -> Int
numTrue lst = length (filter (== True) lst)

movingSum k lst = map sumup $ tails lst
  where
    sumup = sum . take k

day01a :: String -> Int
day01a = numTrue . hasIncreased . strToIntList

day01b :: String -> Int
day01b = numTrue . hasIncreased . movingSum 3 . strToIntList

File: python/adventofcode/Year2021/Day01.py
def day01a(numbers: list[str]) -> int:
    this = None
    increases = 0

    for number in numbers:
        last, this = this, int(number)
        if last is None:
            continue

        if last < this:
            increases += 1
    return increases


def day01b(numbers: list[str]) -> int:
    k: int = 3
    this: list[int] = []
    increases = 0

    for number in numbers:
        last, this = this, this + [int(number)]

        if len(last) < k:
            continue

        this.pop(0)
        if sum(last) < sum(this):
            increases += 1
    return increases
